# SeqWord Motif Mapper v3.2.6 - Help Guide

## General Usage

`python run.py [-arguments]`

`python run.py -h / -H / --help` - Show this help

`python run.py -v / -V / --version` - Show version

This program processes GFF files generated by the program ipdSummary and generates graphical representations, along with performing statistical analysis of modified base distributions.

## Example Usage

`python program.py -i input.gff -g genome.gbk -d example -mm Y -w GATC,2,-2 -p 100 -sp Y`

## General Settings

| Short Option | Long Option | Type | Default | Description |
| --- | --- | --- | --- | --- |
| \-d | \--project\_directory | path | _optional_ | Project subfolder in 'input' folder |
| \-i | \--input\_GFF | path | _mandatory_ | Input GFF file name |
| \-g | \--input\_GBK | path | _mandatory_ | Genome GBK file |
| \-m | \--filter\_file | path | _optional_ | Filter regions |
| \-ft | \--input\_GBK | string | _optional_ | Generic file title |
| \-ogf | \--output\_graph\_format | string | HTML | Output graph format (SVG, HTML, PDF, EPS, JPG, PNG, BMP) |
| \-dpi | \--dpi | integer | 300 | Raster DPI (between 100 and 1200 dpi) |
| \-p | \--promoter\_length | integer | 75  | Promoter region length |
| \-r | \--maximum\_sites | integer | 10000 | Maximum number of sites for verification (0 to skip) |
| \-n | \--blast\_context\_mismatch | integer | 2   | Allowed number of context mismatches |
| \-z | \--blast\_motif\_mismatch | Yes/No | Y   | Allow motif mismatch |
| _Options not recommended for changing:_ |     |     |     |     |
| \-u | \--input\_folder | path | input | Input folder |
| \-o | \--output\_folder | path | output | Output folder |
| \-tmp | \--tmp\_folder | path | ./lib/bin/tmp | TMP folder |
| \-x | \--bin\_folder | path | ./lib/bin | Executable folder |

## Circular Map Settings

| Short Option | Long Option | Type | Default | Description |
| --- | --- | --- | --- | --- |
| \-mm | \--circular\_map | Y/N | Y   | Generate circular plot graph, Y provides access to the following options: |
| \-w | \--cmap\_motif | string | _mandatory_ | Motif (e.g., GATC,2,-2) |
| \-s | \--sites\_or\_motifs | S/M | sites | Search for sites (S) or motifs (M) |
| \-f | \--modified\_or\_unmodified | M/U | M   | Modified (M) or Unmodified (U) motifs |
| \-wl | \--window\_length | integer | 8000 | Sliding window length |
| \-ws | \--window\_step | integer | 2000 | Sliding window step |
| \-c | \--cmap\_score\_cutoff | integer | 21  | Circular map score cutoff (0 to skip) |
| \-cmt | \--cmap\_graph\_title | string | _optional_ | Circular map graph title |

## Dot-Plot Settings

| Short Option | Long Option | Type | Default | Description |
| --- | --- | --- | --- | --- |
| \-dp | \--dotplot | Y/N | N   | Generate dot-plot graph, Y provides access to the following options: |
| \-dpn | \--nucleotides | A,C,G,T | A,C | Modified nucleotides to display (void to skip) |
| \-dpm | \--methylation\_types | m6A,m4C | _optional_ | Methylation types (void to skip) |
| \-dpf | \--dotplot\_motifs | string | _optional_ | Methylation motifs to include/exclude (e.g., GATC,2,-2; -CRGKGATC,1,6,-2) |
| \-dpc | \--dotplot\_score\_cutoff | integer | 21  | Dotplot score cutoff (0 to skip) |
| \-dpw | \--maximum\_coverage | integer | 0   | Maximum coverage (X value) to plot (0 to skip) |
| \-dps | \--maximum\_score | integer | 0   | Maximum score (Y value) to plot (0 to skip) |
| \-dpt | \--dp\_graph\_title | string | _optional_ | Dot-plot graph title |

## Statistics Panel Settings

| Short Option | Long Option | Type | Default | Description |
| --- | --- | --- | --- | --- |
| \-sp | \--statplot | Y/N | Y   | Generate statistics panel, Y provides access to the following options: |
| \-tsk | \--tasks | string | gc,gcs | Tasks like **gc** (GC-skew), **gcs** (GC-skew), **mge** (identify mobile genetic elements) |
| \-std | \--strand | string | off | Exclude strand (leading, lagging, off) or (DIR \| REV \| BOTH) or (1, -1, 0) |
| \-spt | \--sp\_graph\_title | string | _optional_ | Statistical plot graph title |

## Contact

For further information, refer to the official documentation or contact [oleg.reva@up.ac.za](mailto:oleg.reva@up.ac.za)
