<h1>SeqWord Motif Mapper (SWMM) v3.2.6 - Help Guide</h1>
  
  <h2>General Usage</h2>
  <p>This program processes GFF files generated by the ipdSummary program and generates different plots to view bacterial methylation patterns. This is achieved by performing statistical analyses on modified base distributions. </p>
  <p>To run the program, use the web application at <a href='http://begp.bi.up.ac.za/'>http://begp.bi.up.ac.za/</a> or download the command line tool described below.</p>
  <p> Worked examples can be found at <a href='https://begp.bi.up.ac.za/usage.html'>https://begp.bi.up.ac.za/usage.html</a> and in the paper at https://papers.ssrn.com/sol3/papers.cfm?abstract_id=5232247. </p> 
  <p> If you have found SWMM to be useful for your studies, please remember to cite the paper. </p>
  <p><code>python run.py [-arguments]</code></p>
  <p><code>python run.py -h / -H / --help</code> - Show this help</p>
  <p><code>python run.py -v / -V / --version</code> - Show version</p>


  <h2>Example Usage</h2>
  <p><code>python program.py -i input.gff -g genome.gbk -d example -mm Y -w GATC,2,-2 -p 100 -sp Y</code></p>
  
  <h2>General Settings</h2>
  <table>
      <tr>
          <th>Short Option</th>
          <th>Long Option</th>
          <th>Type</th>
          <th>Default</th>
          <th>Description</th>
      </tr>
      <tr><td>-d</td><td>--project_directory</td><td>path</td><td><i>optional</i></td><td>Project subfolder in 'input' folder</td></tr>
      <tr><td>-i</td><td>--input_GFF</td><td>path</td><td><i>mandatory</i></td><td>Input GFF file name</td></tr>
      <tr><td>-g</td><td>--input_GBK</td><td>path</td><td><i>mandatory</i></td><td>Genome GBK file</td></tr>
      <tr><td>-m</td><td>--filter_file</td><td>path</td><td><i>optional</i></td><td>Filter regions</td></tr>
      <tr><td>-ft</td><td>--input_GBK</td><td>string</td><td><i>optional</i></td><td>Generic file title</td></tr>
      <tr><td>-ogf</td><td>--output_graph_format</td><td>string</td><td>HTML</td><td>Output graph format (SVG, HTML, PDF, EPS, JPG, PNG, BMP)</td></tr>
      <tr><td>-dpi</td><td>--dpi</td><td>integer</td><td>300</td><td>Raster DPI (between 100 and 1200 dpi)</td></tr>
      <tr><td>-p</td><td>--promoter_length</td><td>integer</td><td>75</td><td>Promoter region length</td></tr>
      <tr><td>-r</td><td>--maximum_sites</td><td>integer</td><td>10000</td><td>Maximum number of sites for verification (0 to skip)</td></tr>
      <tr><td>-n</td><td>--blast_context_mismatch</td><td>integer</td><td>2</td><td>Allowed number of context mismatches</td></tr>
      <tr><td>-z</td><td>--blast_motif_mismatch</td><td>Yes/No</td><td>Y</td><td>Allow motif mismatch</td></tr>
      <tr><td colspan=5><i>Options not recommended for changing:</i></td></tr>
      <tr><td>-u</td><td>--input_folder</td><td>path</td><td>input</td><td>Input folder</td></tr>
      <tr><td>-o</td><td>--output_folder</td><td>path</td><td>output</td><td>Output folder</td></tr>
      <tr><td>-tmp</td><td>--tmp_folder</td><td>path</td><td>./lib/bin/tmp</td><td>TMP folder</td></tr>
      <tr><td>-x</td><td>--bin_folder</td><td>path</td><td>./lib/bin</td><td>Executable folder</td></tr>
  </table>
  
  <h2>Circular Map Settings</h2>
  <table>
      <tr>
          <th>Short Option</th>
          <th >Long Option</th>
          <th>Type</th>
          <th>Default</th>
          <th>Description</th>
      </tr>
      <tr><td>-mm</td><td>--circular_map</td><td>Y/N</td><td>Y</td><td>Generate circular plot graph, Y provides access to the following options:</td></tr>
      <tr><td>-w</td><td>--cmap_motif</td><td>string</td><td><i>mandatory</i></td><td>Motif (e.g., GATC,2,-2)</td></tr>
      <tr><td>-s</td><td>--sites_or_motifs</td><td>S/M</td><td>sites</td><td>Search for sites (S) or motifs (M)</td></tr>
      <tr><td>-f</td><td>--modified_or_unmodified</td><td>M/U</td><td>M</td><td>Modified (M) or Unmodified (U) motifs</td></tr>
      <tr><td>-wl</td><td>--window_length</td><td>integer</td><td>8000</td><td>Sliding window length</td></tr>
      <tr><td>-ws</td><td>--window_step</td><td>integer</td><td>2000</td><td>Sliding window step</td></tr>
      <tr><td>-c</td><td>--cmap_score_cutoff</td><td>integer</td><td>21</td><td>Circular map score cutoff (0 to skip)</td></tr>
      <tr><td>-cmt</td><td>--cmap_graph_title</td><td>string</td><td><i>optional</i></td><td>Circular map graph title</td></tr>
  </table>
  
  <h2>Dot-Plot Settings</h2>
  <table>
      <tr>
          <th>Short Option</th>
          <th>Long Option</th>
          <th>Type</th>
          <th>Default</th>
          <th>Description</th>
      </tr>
      <tr><td>-dp</td><td>--dotplot</td><td>Y/N</td><td>N</td><td>Generate dot-plot graph, Y provides access to the following options:</td></tr>
      <tr><td>-dpn</td><td>--nucleotides</td><td>A,C,G,T</td><td>A,C</td><td>Modified nucleotides to display (void to skip)</td></tr>
      <tr><td>-dpm</td><td>--methylation_types</td><td>m6A,m4C</td><td><i>optional</i></td><td>Methylation types (void to skip)</td></tr>
      <tr><td>-dpf</td><td>--dotplot_motifs</td><td>string</td><td><i>optional</i></td><td>Methylation motifs to include/exclude (e.g., GATC,2,-2; -CRGKGATC,1,6,-2)</td></tr>
      <tr><td>-dpc</td><td>--dotplot_score_cutoff</td><td>integer</td><td>21</td><td>Dotplot score cutoff (0 to skip)</td></tr>
      <tr><td>-dpw</td><td>--maximum_coverage</td><td>integer</td><td>0</td><td>Maximum coverage (X value) to plot (0 to skip)</td></tr>
      <tr><td>-dps</td><td>--maximum_score</td><td>integer</td><td>0</td><td>Maximum score (Y value) to plot (0 to skip)</td></tr>
      <tr><td>-dpt</td><td>--dp_graph_title</td><td>string</td><td><i>optional</i></td><td>Dot-plot graph title</td></tr>
  </table>
  
  <h2>Statistics Panel Settings</h2>
  <table>
      <tr>
          <th>Short Option</th>
          <th>Long Option</th>
          <th>Type</th>
          <th>Default</th>
          <th>Description</th>
      </tr>
      <tr><td>-sp</td><td>--statplot</td><td>Y/N</td><td>Y</td><td>Generate statistics panel, Y provides access to the following options:</td></tr>
      <tr><td>-tsk</td><td>--tasks</td><td>string</td><td>gc,gcs</td><td>Tasks like <b>gc</b> (GC-skew), <b>gcs</b> (GC-skew), <b>mge</b> (identify mobile genetic elements)</td></tr>
      <tr><td>-std</td><td>--strand</td><td>string</td><td>off</td><td>Exclude strand (leading, lagging, off) or (DIR | REV | BOTH) or (1, -1, 0)</td></tr>
      <tr><td>-spt</td><td>--sp_graph_title</td><td>string</td><td><i>optional</i></td><td>Statistical plot graph title</td></tr>
  </table>
  
  <h2>Contact</h2>
  <p>For further information, refer to the official documentation or contact <a href="mailto:oleg.reva@up.ac.za">oleg.reva@up.ac.za</a></p>


